File: iOS/Daily/Views/Health/MonkeyMascotView.swift
import SwiftUI
import DailyCore

/// Size variant for the stylized monkey mascot.
public enum MonkeySize {
    case badge // 24x24 for pills and widgets
    case mini  // 44x44 for widget headers and list items
    case card  // 80x80 for dashboard cards
    case hero  // 130x130 for Stress Studio hero
    
    var dimension: CGFloat {
        switch self {
        case .badge: return 24
        case .mini: return 44
        case .card: return 80
        case .hero: return 130
        }
    }
}

/// A charming, procedural vector monkey mascot with expressive moods and animated autonomic breath glow.
public struct MonkeyMascotView: View {
    public let mood: MonkeyMood
    public let size: MonkeySize
    public let animated: Bool
    
    @State private var breatheScale: CGFloat = 1.0
    @State private var earWiggle: Double = 0.0
    
    public init(mood: MonkeyMood = .curious, size: MonkeySize = .card, animated: Bool = true) {
        self.mood = mood
        self.size = size
        self.animated = animated
    }
    
    private var primaryAuraColor: Color {
        switch mood {
        case .zen: return Color(hex: "#00E5FF")        // Mint / neon cyan
        case .curious: return Color(hex: "#00FFB2")    // Teal / spring green
        case .busy: return Color(hex: "#FFA726")       // Warm amber
        case .overheated: return Color(hex: "#FF5252") // Coral red
        }
    }
    
    public var body: some View {
        let dim = size.dimension
        
        ZStack {
            // Background Autonomic Aura Halo
            Circle()
                .fill(
                    RadialGradient(
                        colors: [
                            primaryAuraColor.opacity(0.35),
                            primaryAuraColor.opacity(0.08),
                            Color.clear
                        ],
                        center: .center,
                        startRadius: dim * 0.2,
                        endRadius: dim * 0.65
                    )
                )
                .scaleEffect(breatheScale)
            
            // Stylized Monkey Head & Facial Features
            monkeyFace(dimension: dim)
        }
        .frame(width: dim, height: dim)
        .onAppear {
            if animated {
                withAnimation(
                    .easeInOut(duration: mood == .overheated ? 1.4 : 3.0)
                    .repeatForever(autoreverses: true)
                ) {
                    breatheScale = 1.12
                    earWiggle = mood == .busy ? 4.0 : 1.5
                }
            }
        }
    }
    
    @ViewBuilder
    private func monkeyFace(dimension: CGFloat) -> some View {
        let scale = dimension / 100.0
        
        ZStack {
            // Ears (Outer)
            HStack(spacing: 64 * scale) {
                // Left Ear
                earView(scale: scale, isLeft: true)
                // Right Ear
                earView(scale: scale, isLeft: false)
            }
            .offset(y: -6 * scale)
            
            // Main Head (Warm Caramel Coat)
            Circle()
                .fill(
                    LinearGradient(
                        colors: [
                            Color(hex: "#8D5B4C"),
                            Color(hex: "#6D3B2C")
                        ],
                        startPoint: .top,
                        endPoint: .bottom
                    )
                )
                .frame(width: 72 * scale, height: 68 * scale)
                .overlay(
                    Circle()
                        .stroke(primaryAuraColor.opacity(0.4), lineWidth: 1.5 * scale)
                )
            
            // Face Mask (Peaches & Cream Inset)
            HStack(spacing: -8 * scale) {
                Circle()
                    .fill(Color(hex: "#F7D6BF"))
                    .frame(width: 36 * scale, height: 36 * scale)
                Circle()
                    .fill(Color(hex: "#F7D6BF"))
                    .frame(width: 36 * scale, height: 36 * scale)
            }
            .offset(y: -4 * scale)
            
            // Lower Muzzle
            Capsule()
                .fill(Color(hex: "#F7D6BF"))
                .frame(width: 50 * scale, height: 32 * scale)
                .offset(y: 10 * scale)
            
            // Eyes & Expression
            eyesView(scale: scale)
                .offset(y: -2 * scale)
            
            // Nose
            Capsule()
                .fill(Color(hex: "#4E271E"))
                .frame(width: 8 * scale, height: 5 * scale)
                .offset(y: 9 * scale)
            
            // Mouth Expression
            mouthView(scale: scale)
                .offset(y: 18 * scale)
            
            // Mood Specific Accessories
            moodAccessory(scale: scale)
        }
    }
    
    // Ears with Inner Pink Cavity
    @ViewBuilder
    private func earView(scale: CGFloat, isLeft: Bool) -> some View {
        ZStack {
            Circle()
                .fill(Color(hex: "#7A4333"))
                .frame(width: 24 * scale, height: 24 * scale)
            Circle()
                .fill(Color(hex: "#F7B8A1"))
                .frame(width: 14 * scale, height: 14 * scale)
        }
        .rotationEffect(.degrees(isLeft ? -earWiggle : earWiggle))
    }
    
    // Expressive Eyes Based on Stress State
    @ViewBuilder
    private func eyesView(scale: CGFloat) -> some View {
        HStack(spacing: 16 * scale) {
            switch mood {
            case .zen:
                // Serene curved closed eyes (smiling arcs)
                Text("◜  ◝")
                    .font(.system(size: 14 * scale, weight: .bold, design: .rounded))
                    .foregroundColor(Color(hex: "#3D1E15"))
                
            case .curious:
                // Big round shiny eyes with light reflection
                eyePupil(scale: scale)
                eyePupil(scale: scale)
                
            case .busy:
                // Thoughtful focused eyes (one slightly squinted)
                eyePupil(scale: scale)
                Circle()
                    .fill(Color(hex: "#2D120B"))
                    .frame(width: 6.5 * scale, height: 5 * scale)
                
            case .overheated:
                // Squinting stress eyes: > <
                HStack(spacing: 14 * scale) {
                    Text("✕")
                        .font(.system(size: 10 * scale, weight: .heavy))
                        .foregroundColor(Color(hex: "#FF5252"))
                    Text("✕")
                        .font(.system(size: 10 * scale, weight: .heavy))
                        .foregroundColor(Color(hex: "#FF5252"))
                }
            }
        }
    }
    
    @ViewBuilder
    private func eyePupil(scale: CGFloat) -> some View {
        ZStack(alignment: .topTrailing) {
            Circle()
                .fill(Color(hex: "#241009"))
                .frame(width: 8 * scale, height: 8 * scale)
            Circle()
                .fill(Color.white)
                .frame(width: 2.5 * scale, height: 2.5 * scale)
                .offset(x: -1.5 * scale, y: 1.5 * scale)
        }
    }
    
    // Expressive Mouth
    @ViewBuilder
    private func mouthView(scale: CGFloat) -> some View {
        switch mood {
        case .zen:
            // Gentle serene smile
            Text("‿")
                .font(.system(size: 14 * scale, weight: .semibold, design: .rounded))
                .foregroundColor(Color(hex: "#4E271E"))
                .offset(y: -4 * scale)
            
        case .curious:
            // Cute open half-smile
            Capsule()
                .fill(Color(hex: "#D85A5A"))
                .frame(width: 8 * scale, height: 4 * scale)
            
        case .busy:
            // Straight pensive line
            Capsule()
                .fill(Color(hex: "#4E271E"))
                .frame(width: 10 * scale, height: 2 * scale)
            
        case .overheated:
            // Wavy or open sigh mouth
            Capsule()
                .fill(Color(hex: "#C62828"))
                .frame(width: 10 * scale, height: 6 * scale)
        }
    }
    
    // Mood Specific Accessories (Zen Lotus / Overheated Ice Pack)
    @ViewBuilder
    private func moodAccessory(scale: CGFloat) -> some View {
        switch mood {
        case .zen:
            // Floating green lotus / leaf
            Image(systemName: "leaf.fill")
                .font(.system(size: 10 * scale))
                .foregroundColor(Color(hex: "#00E5FF"))
                .offset(x: 18 * scale, y: -26 * scale)
                .shadow(color: Color(hex: "#00E5FF").opacity(0.6), radius: 4 * scale)
            
        case .curious:
            EmptyView()
            
        case .busy:
            // Thinking bubble
            Circle()
                .fill(Color.white.opacity(0.85))
                .frame(width: 5 * scale, height: 5 * scale)
                .offset(x: 26 * scale, y: -22 * scale)
            
        case .overheated:
            // Cute cool ice-pack on forehead
            ZStack {
                Capsule()
                    .fill(
                        LinearGradient(
                            colors: [Color(hex: "#64B5F6"), Color(hex: "#1E88E5")],
                            startPoint: .top,
                            endPoint: .bottom
                        )
                    )
                    .frame(width: 28 * scale, height: 12 * scale)
                Text("❄︎")
                    .font(.system(size: 8 * scale, weight: .bold))
                    .foregroundColor(.white)
            }
            .offset(y: -28 * scale)
            .shadow(color: Color.black.opacity(0.3), radius: 3 * scale)
        }
    }
}

File: Android/core-designsystem/src/main/java/com/intellidream/daily/designsystem/MonkeyMascotView.kt
package com.intellidream.daily.designsystem

import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Eco
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.intellidream.daily.model.MonkeyMood

/**
 * Size variants for the stylized procedural vector monkey mascot.
 * 1:1 parity with iOS MonkeySize.
 */
enum class MonkeySize(val dimension: Dp) {
    BADGE(24.dp), // For pills and widgets
    MINI(44.dp),  // For widget headers and list items
    CARD(80.dp),  // For dashboard cards
    HERO(130.dp); // For Stress Studio hero

    val scaleFactor: Float
        get() = dimension.value / 100f
}

/**
 * A charming, procedural vector monkey mascot with expressive moods and animated autonomic breath glow.
 * 1:1 Kotlin/Compose port of iOS [MonkeyMascotView.swift].
 */
@Composable
fun MonkeyMascotView(
    mood: MonkeyMood = MonkeyMood.CURIOUS,
    size: MonkeySize = MonkeySize.CARD,
    animated: Boolean = true,
    modifier: Modifier = Modifier
) {
    val auraColor = when (mood) {
        MonkeyMood.ZEN -> Color(0xFF00E5FF)        // Mint / neon cyan
        MonkeyMood.CURIOUS -> Color(0xFF00FFB2)    // Teal / spring green
        MonkeyMood.BUSY -> Color(0xFFFFA726)       // Warm amber
        MonkeyMood.OVERHEATED -> Color(0xFFFF5252) // Coral red
    }

    val animDuration = if (mood == MonkeyMood.OVERHEATED) 1400 else 3000
    val infiniteTransition = rememberInfiniteTransition(label = "monkeyMascotBreath")

    val breatheScale by if (animated) {
        infiniteTransition.animateFloat(
            initialValue = 1.0f,
            targetValue = 1.12f,
            animationSpec = infiniteRepeatable(
                animation = tween(durationMillis = animDuration, easing = FastOutSlowInEasing),
                repeatMode = RepeatMode.Reverse
            ),
            label = "breatheScale"
        )
    } else {
        androidx.compose.runtime.remember { androidx.compose.runtime.mutableFloatStateOf(1.0f) }
    }

    val earWiggle by if (animated) {
        val targetDeg = if (mood == MonkeyMood.BUSY) 4.0f else 1.5f
        infiniteTransition.animateFloat(
            initialValue = -targetDeg,
            targetValue = targetDeg,
            animationSpec = infiniteRepeatable(
                animation = tween(durationMillis = animDuration, easing = FastOutSlowInEasing),
                repeatMode = RepeatMode.Reverse
            ),
            label = "earWiggle"
        )
    } else {
        androidx.compose.runtime.remember { androidx.compose.runtime.mutableFloatStateOf(0.0f) }
    }

    val dim = size.dimension
    val scale = size.scaleFactor

    Box(
        modifier = modifier.size(dim),
        contentAlignment = Alignment.Center
    ) {
        // Background Autonomic Aura Halo
        Canvas(
            modifier = Modifier
                .size(dim)
                .scale(breatheScale)
        ) {
            drawCircle(
                brush = Brush.radialGradient(
                    colors = listOf(
                        auraColor.copy(alpha = 0.35f),
                        auraColor.copy(alpha = 0.08f),
                        Color.Transparent
                    ),
                    center = center,
                    radius = this.size.minDimension * 0.65f
                )
            )
        }

        // Monkey Facial Structure
        Box(
            modifier = Modifier.size(dim),
            contentAlignment = Alignment.Center
        ) {
            // Ears
            Row(
                modifier = Modifier
                    .offset(y = (-6 * scale).dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                // Left Ear
                EarView(scale = scale, isLeft = true, earWiggle = earWiggle)

                Box(modifier = Modifier.size((40 * scale).dp))

                // Right Ear
                EarView(scale = scale, isLeft = false, earWiggle = earWiggle)
            }

            // Main Head (Warm Caramel Coat)
            Box(
                modifier = Modifier
                    .size((72 * scale).dp, (68 * scale).dp)
                    .clip(CircleShape)
                    .background(
                        brush = Brush.verticalGradient(
                            listOf(Color(0xFF8D5B4C), Color(0xFF6D3B2C))
                        )
                    )
                    .border((1.5f * scale).dp, auraColor.copy(alpha = 0.4f), CircleShape)
            )

            // Face Mask (Peaches & Cream Inset)
            Row(
                modifier = Modifier.offset(y = (-4 * scale).dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Box(
                    modifier = Modifier
                        .size((36 * scale).dp)
                        .clip(CircleShape)
                        .background(Color(0xFFF7D6BF))
                )
                Box(
                    modifier = Modifier
                        .offset(x = (-8 * scale).dp)
                        .size((36 * scale).dp)
                        .clip(CircleShape)
                        .background(Color(0xFFF7D6BF))
                )
            }

            // Lower Muzzle
            Box(
                modifier = Modifier
                    .offset(y = (10 * scale).dp)
                    .size((50 * scale).dp, (32 * scale).dp)
                    .clip(CircleShape)
                    .background(Color(0xFFF7D6BF))
            )

            // Eyes & Expression
            EyesView(mood = mood, scale = scale)

            // Nose
            Box(
                modifier = Modifier
                    .offset(y = (9 * scale).dp)
                    .size((8 * scale).dp, (5 * scale).dp)
                    .clip(CircleShape)
                    .background(Color(0xFF4E271E))
            )

            // Mouth
            MouthView(mood = mood, scale = scale)

            // Mood Specific Accessories
            MoodAccessory(mood = mood, scale = scale)
        }
    }
}

@Composable
private fun EarView(scale: Float, isLeft: Boolean, earWiggle: Float) {
    Box(
        modifier = Modifier
            .rotate(if (isLeft) -earWiggle else earWiggle)
            .size((24 * scale).dp)
            .clip(CircleShape)
            .background(Color(0xFF7A4333)),
        contentAlignment = Alignment.Center
    ) {
        Box(
            modifier = Modifier
                .size((14 * scale).dp)
                .clip(CircleShape)
                .background(Color(0xFFF7B8A1))
        )
    }
}

@Composable
private fun EyesView(mood: MonkeyMood, scale: Float) {
    val spacing = (16 * scale).dp

    Box(
        modifier = Modifier.offset(y = (-2 * scale).dp),
        contentAlignment = Alignment.Center
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            when (mood) {
                MonkeyMood.ZEN -> {
                    // Serene curved closed eyes (smiling arcs)
                    Text(
                        text = "◜  ◝",
                        color = Color(0xFF3D1E15),
                        fontSize = (14 * scale).sp,
                        fontWeight = FontWeight.Bold
                    )
                }

                MonkeyMood.CURIOUS -> {
                    EyePupil(scale = scale)
                    Box(modifier = Modifier.size(spacing))
                    EyePupil(scale = scale)
                }

                MonkeyMood.BUSY -> {
                    EyePupil(scale = scale)
                    Box(modifier = Modifier.size(spacing))
                    Box(
                        modifier = Modifier
                            .size((6.5f * scale).dp, (5 * scale).dp)
                            .clip(CircleShape)
                            .background(Color(0xFF2D120B))
                    )
                }

                MonkeyMood.OVERHEATED -> {
                    Text(
                        text = "✕",
                        color = Color(0xFFFF5252),
                        fontSize = (10 * scale).sp,
                        fontWeight = FontWeight.Black
                    )
                    Box(modifier = Modifier.size((14 * scale).dp))
                    Text(
                        text = "✕",
                        color = Color(0xFFFF5252),
                        fontSize = (10 * scale).sp,
                        fontWeight = FontWeight.Black
                    )
                }
            }
        }
    }
}

@Composable
private fun EyePupil(scale: Float) {
    Box(
        modifier = Modifier
            .size((8 * scale).dp)
            .clip(CircleShape)
            .background(Color(0xFF241009))
    ) {
        Box(
            modifier = Modifier
                .offset(x = (1.5f * scale).dp, y = (1.5f * scale).dp)
                .size((2.5f * scale).dp)
                .clip(CircleShape)
                .background(Color.White)
        )
    }
}

@Composable
private fun MouthView(mood: MonkeyMood, scale: Float) {
    Box(
        modifier = Modifier.offset(y = (18 * scale).dp),
        contentAlignment = Alignment.Center
    ) {
        when (mood) {
            MonkeyMood.ZEN -> {
                Text(
                    text = "‿",
                    color = Color(0xFF4E271E),
                    fontSize = (14 * scale).sp,
                    fontWeight = FontWeight.SemiBold,
                    modifier = Modifier.offset(y = (-4 * scale).dp)
                )
            }

            MonkeyMood.CURIOUS -> {
                Box(
                    modifier = Modifier
                        .size((8 * scale).dp, (4 * scale).dp)
                        .clip(CircleShape)
                        .background(Color(0xFFD85A5A))
                )
            }

            MonkeyMood.BUSY -> {
                Box(
                    modifier = Modifier
                        .size((10 * scale).dp, (2 * scale).dp)
                        .clip(CircleShape)
                        .background(Color(0xFF4E271E))
                )
            }

            MonkeyMood.OVERHEATED -> {
                Box(
                    modifier = Modifier
                        .size((10 * scale).dp, (6 * scale).dp)
                        .clip(CircleShape)
                        .background(Color(0xFFC62828))
                )
            }
        }
    }
}

@Composable
private fun MoodAccessory(mood: MonkeyMood, scale: Float) {
    when (mood) {
        MonkeyMood.ZEN -> {
            // Floating green lotus / leaf
            Icon(
                imageVector = Icons.Rounded.Eco,
                contentDescription = null,
                tint = Color(0xFF00E5FF),
                modifier = Modifier
                    .offset(x = (18 * scale).dp, y = (-26 * scale).dp)
                    .size((12 * scale).dp)
            )
        }

        MonkeyMood.CURIOUS -> {
            // Clean curious look
        }

        MonkeyMood.BUSY -> {
            // Thinking bubble
            Box(
                modifier = Modifier
                    .offset(x = (26 * scale).dp, y = (-22 * scale).dp)
                    .size((5 * scale).dp)
                    .clip(CircleShape)
                    .background(Color.White.copy(alpha = 0.85f))
            )
        }

        MonkeyMood.OVERHEATED -> {
            // Cute cool ice-pack on forehead
            Box(
                modifier = Modifier
                    .offset(y = (-28 * scale).dp)
                    .size((28 * scale).dp, (12 * scale).dp)
                    .shadow((3 * scale).dp, CircleShape)
                    .clip(CircleShape)
                    .background(
                        brush = Brush.verticalGradient(
                            listOf(Color(0xFF64B5F6), Color(0xFF1E88E5))
                        )
                    ),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "❄",
                    color = Color.White,
                    fontSize = (8 * scale).sp,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}
